File: ruby/lib/browserstack-local.rb
require 'net/http'
require 'rbconfig'

class BrowserStackLocal
  attr_reader :pid

  def initialize(key, binary_path = nil)
    @key = key
    @binary_path = if binary_path.nil?
        LocalBinary.new.download
      else
        binary_path
      end
  end

  def add_args(key, value=nil)
    if key == "v"
      @verbose_flag = "-v"
    elsif key == "force"
      @force_flag = "-force"
    elsif key == "only"
      @only_flag = "-only"
    elsif key == "onlyAutomate"
      @only_automate_flag = "-onlyAutomate"
    elsif key == "forcelocal"
      @force_local_flag = "-forcelocal"
    elsif key == "localIdentifier"
      @local_identifier_flag = "-localIdentifier '#{value}'"
    elsif key == "f"
      @folder_flag = "-f"
      @folder_path = "'#{value}'"
    elsif key == "proxyHost"
      @proxy_host = "-proxyHost '#{value}'"
    elsif key == "proxyPort"
      @proxy_port = "-proxyPort #{value}"
    elsif key == "proxyUser"
      @proxy_user = "-proxyUser '#{value}'"
    elsif key == "proxyPass"
      @proxy_pass = "-proxyPass '#{value}'"
    elsif key == "hosts"
      @hosts = value
    end
  end

  def start
    @process = IO.popen(command, "w+")

    while true
      line = @process.readline
      break if line.nil?
      if line.match(/\*\*\* Error\:/)
        @process.close
        raise BrowserStackLocalException.new(line)
        return
      end
      if line.strip == "Press Ctrl-C to exit"
        @pid = @process.pid
        return
      end
    end

    while true
      resp = Net::HTTP.get(URI.parse("http://localhost:45691/check")) rescue nil
      puts resp
      break if resp && resp.match(/running/i)
      sleep 1
    end
  end

  def stop
    return if @pid.nil?
    Process.kill("INT", @pid)
    @process.close
  end

  def command
    "#{@binary_path} #{@folder_flag} #{@key} #{@folder_path} #{@force_local_flag} #{@local_identifier_flag} #{@only_flag} #{@only_automate_flag} #{@proxy_host} #{@proxy_port} #{@proxy_user} #{@proxy_pass} #{@force_flag} #{@verbose_flag} #{@hosts}".strip
  end
end

class BrowserStackLocalException < Exception
end

class LocalBinary
  def initialize
    host_os = RbConfig::CONFIG['host_os']
    @http_path = case host_os
    when /mswin|msys|mingw|cygwin|bccwin|wince|emc/
      @windows = true
      "https://s3.amazonaws.com/browserStack/browserstack-local/BrowserStackLocal.exe"
    when /darwin|mac os/
      "https://s3.amazonaws.com/browserStack/browserstack-local/BrowserStackLocal-darwin-x64"
    when /linux/
      if 1.size == 8
        "https://s3.amazonaws.com/browserStack/browserstack-local/BrowserStackLocal-linux-x64"
      else
        "https://s3.amazonaws.com/browserStack/browserstack-local/BrowserStackLocal-linux-ia32"
      end
    end
  end

  def download
    dest_parent_dir = File.join(File.expand_path('~'), '.browserstack')
    unless File.exists? dest_parent_dir
      Dir.mkdir dest_parent_dir
    end
    res = Net::HTTP.get(@http_path, :use_ssl => true)
    binary_path = File.join(dest_parent_dir, "BrowserStackLocal#{".exe" if @windows}")
    open(binary_path) do |file|
      file.write(res.body)
    end
    binary_path
  end
end

File: php/lib/BrowserStackLocal.php
<?php

namespace BrowserStack;

use Exception;

error_reporting(1);

class BrowserStackLocal {

  private $handle = NULL;
  private $pipes = array();

	public function __construct($key) {
    $this->key = $key;
    
    if (!is_executable("BrowserStack/BrowserStackLocal"))
      $this->prepare_binary();
  }

	public function __destruct() {
        echo "";
  }

 	public function verbose() {
    	$this->verbose_flag = "-v";
  }

	public function enable_folder($path = "") {
  	$this->folder_flag = "-f";
  	$this->folder_path = $path;
	}

  public function enable_force() {
    $this->force_flag = "-force";
  }

  public function enable_only() {
    $this->only_flag = "-only";
  }

  public function enable_only_automate() {
    $this->only_automate_flag = "-onlyAutomate";
  }

  public function enable_force_local() {
    $this->force_local_flag = "-forcelocal";
  }

  public function set_local_identifier($localIdentifier = "") {
    $this->local_identifier_flag = "-localIdentifier $localIdentifier";
  }

  public function start() {

    $descriptorspec = array(
      0 => array("pipe", "r"),  // stdin is a pipe that the child will read from
      1 => array("pipe", "w"),  // stdout is a pipe that the child will write to
      2 => array("file", "/tmp/error-output.txt", "a") // stderr is a file to write to
    );
    
    $call = $this->command();
    
    $this->handle = proc_open($call, $descriptorspec,$this->pipes);
  
    while(!feof($this->pipes[1])) {
        $buffer = fgets($this->pipes[1]);
        if (preg_match("/\bError\b/i", $buffer,$match)) {
          throw new BrowserStackLocalException($buffer);
          proc_terminate($this->handle);
          return;
        }
        elseif (strcmp(rtrim($buffer),"Press Ctrl-C to exit") == 0)
          return;
        flush();    
    }

  }

  public function stop() {
    if (is_null($this->handle))
      return;
    else {
      echo "hello";
      proc_terminate($this->handle);
    }
  }

private function platform_url()
{
  if (PHP_OS == "Darwin")
    return "https://www.browserstack.com/browserstack-local/BrowserStackLocal-darwin-x64.zip";
  else if (strtoupper(substr(PHP_OS, 0, 3)) === 'WIN')
    return "https://www.browserstack.com/browserstack-local/BrowserStackLocal-win32.zip";
  if ((strtoupper(PHP_OS)) == "LINUX") {
    if (PHP_INT_SIZE * 8 == 64)
      return "https://www.browserstack.com/browserstack-local/BrowserStackLocal-linux-x64.zip";
    else
      return "https://www.browserstack.com/browserstack-local/BrowserStackLocal-linux-ia32.zip";
  }
}

  public function prepare_binary($url) {
    $url = $this->platform_url();
    
    mkdir('BrowserStack', 0777, true);
    
    file_put_contents("BrowserStack/BrowserStack.zip", fopen($url, 'r'));

    $zip = new \ZipArchive;
    if ($zip->open('BrowserStack/BrowserStack.zip') === TRUE) {
      $zip->extractTo("BrowserStack/");
      $zip->close();
    } else {
      
    }
    chmod("BrowserStack/BrowserStackLocal", 0777);
  }

  public function command() {
    $command = "./BrowserStack/BrowserStackLocal $this->folder_flag $this->key $this->folder_path $this->force_local_flag $this->local_identifier_flag $this->only_flag $this->only_automate_flag $this->force_flag $this->verbose_flag";
    $command = preg_replace('/\s+/S', " ", $command);
    return $command;
  }

  private function is_windows(){
    if(PHP_OS == "WINNT" || PHP_OS == "WIN32"){
        return true;
    }
    return false;
  }
}

